### ISSUE TITLE
Lint when `|` annotations contain string segments

### ISSUE THREAD

**Issue by tylerlaprade** [astral-sh/ruff#9139]
<img width="1138" alt="image" src="https://github.com/astral-sh/ruff/assets/5475199/4fffbbd1-c71b-427e-819c-3d0f5cd67009">

This is one of the inputs to my function, after quotes were automatically added (and I manually turned `QuerySet['ContractVersion']` into `'QuerySet[ContractVersion]'` due to #9135 and #9136:
```
contract_versions_list: list['ContractVersion']
    | 'QuerySet[ContractVersion]'
    | None = None,
```

When I manually changed it to the following, the tests ran fine again:
```
contract_versions_list: 'list[ContractVersion] | QuerySet[ContractVersion]| None' = None,
```

I'm on Ruff v0.1.8 in a Django 4.2.7 project, running tests with PyTest 7.4.3

**Comment by charliermarsh** [comment]
What was the original annotation?

**Comment by charliermarsh** [comment]
The issue here is that you changed `QuerySet['ContractVersion']` to `'QuerySet[ContractVersion]'` which isn't valid -- you need to quote the entire binary operation. But I'm just trying to understand, because you applied that change yourself?

**Comment by tylerlaprade** [comment]
Yes, it looks like I introduced this error when trying to manually resolve the warning from Ruff, since I didn't realize it was invalid to quote just part of it. I'm surprised Pyright doesn't complain about an invalid annotation. Is it possible to make Ruff complain about it?

My original annotation had no quotes in it:
```
list[ContractVersion]
    | QuerySet[ContractVersion]
    | None = None,
```

**Comment by charliermarsh** [comment]
We could probably lint against that, yeah.

**Comment by charliermarsh** [comment]
Repurposing the issue to lint against `"int" | str` which is invalid.
